src/native/libs/System.Security.Cryptography.Native/pal_evp_pkey_ml_dsa.c

```diff
@@ -0,0 +1,237 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "pal_evp_pkey.h"
+#include "pal_evp_pkey_ml_dsa.h"
+#include "pal_utilities.h"
+#include "openssl.h"
+#include <assert.h>
+
+EVP_PKEY* CryptoNative_MLDsaGenerateKey(const char* keyType, uint8_t* seed, int32_t seedLen)
+{
+#if defined(NEED_OPENSSL_3_0) && defined(HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT)
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return NULL;
+    }
+
+    ERR_clear_error();
+
+    if (seed && seedLen != 32)
+    {
+        return NULL;
+    }
+
+    EVP_PKEY_CTX* pctx = EVP_PKEY_CTX_new_from_name(NULL, keyType, NULL);
+    EVP_PKEY* pkey = NULL;
+
+    if (!pctx)
+    {
+        return NULL;
+    }
+
+    if (EVP_PKEY_keygen_init(pctx) <= 0)
+    {
+        goto done;
+    }
+
+    if (seed)
+    {
+        OSSL_PARAM params[] =
+        {
+            OSSL_PARAM_construct_octet_string(OSSL_PKEY_PARAM_ML_DSA_SEED, (void*)seed, Int32ToSizeT(seedLen)),
+            OSSL_PARAM_construct_end(),
+        };
+
+        if (EVP_PKEY_CTX_set_params(pctx, params) <= 0)
+        {
+            goto done;
+        }
+    }
+
+    if (EVP_PKEY_keygen(pctx, &pkey) != 1 && pkey != NULL)
+    {
+        EVP_PKEY_free(pkey);
+        pkey = NULL;
+    }
+
+done:
+    if (pctx != NULL)
+    {
+        EVP_PKEY_CTX_free(pctx);
+    }
+
+    return pkey;
+#else
+    (void)keyType;
+    (void)seed;
+    (void)seedLen;
+    return NULL;
+#endif
+}
+
+int32_t CryptoNative_MLDsaSignPure(EVP_PKEY *pkey,
+                                   void* extraHandle,
+                                   uint8_t* msg, int32_t msgLen,
+                                   uint8_t* context, int32_t contextLen,
+                                   uint8_t* destination, int32_t destinationLen)
+{
+    assert(pkey);
+    assert(msg);
+    assert(msgLen >= 0);
+    assert(contextLen >= 0);
+    assert(destination);
+    assert(destinationLen >= 2420 /* ML-DSA-44 signature size */);
+
+#if defined(NEED_OPENSSL_3_0) && defined(HAVE_OPENSSL_EVP_PKEY_SIGN_MESSAGE_INIT)
+    if (!API_EXISTS(EVP_PKEY_sign_message_init) ||
+        !API_EXISTS(EVP_PKEY_verify_message_init))
+    {
+        return -1;
+    }
+
+    ERR_clear_error();
+
+    EVP_PKEY_CTX* ctx = NULL;
+    EvpPKeyExtraHandle* extra = (EvpPKeyExtraHandle*)extraHandle;
+    OSSL_LIB_CTX* libCtx = extra ? extra->libCtx : NULL;
```

Comment:
I believe these are no longer used in the sign or verify functions (they get handled by EvpPKeyCtxCreateFromPKey) and can be deleted.